core/libp2p/protocol/echo/client_echo_session.cpp

```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "libp2p/protocol/echo/client_echo_session.hpp"
+
+namespace libp2p::protocol {
+
+  ClientEchoSession::ClientEchoSession(
+      std::shared_ptr<connection::Stream> stream)
+      : stream_(std::move(stream)) {
+    BOOST_ASSERT(stream_ != nullptr);
+  }
+
+  void ClientEchoSession::sendAnd(const std::string &send,
+                                  ClientEchoSession::Then then) {
+    if (stream_->isClosedForWrite()) {
+      return;
+    }
+
+    buf_ = std::vector<uint8_t>(send.begin(), send.end());
+
+    BOOST_ASSERT(stream_ != nullptr);
+    stream_->write(
+        buf_, buf_.size(),
+        [this, then{std::move(then)}](outcome::result<size_t> rw) mutable {
```

Comment:
Is it safe to capture `this`? If yes, remove `enable_shared_from_this`, if not - `shared_from_this`